### PR TITLE
compiler: Reuse twin elemental functions

### PR DESCRIPTION
**Pull request by FabioLuporini**
This PR adds machinery to perform the following optimisation for code size (dummy example follows)

Given

```
void foo0(int a, int b):
    return a + b

void foo1(int c, int d):
    return c + d

void bar():
    return foo0(a, b) + foo1(e, f) 
```

Produce

```
void foo0(int a, int b):
    return a + b

void bar():
    return foo0(a, b) + foo0(e, f) 
```

For this, a minor refactoring of the pickling subsystem was required, because now all symbolic objects inherit from an ancestor class `Reconstructable` (which in turn gets used by Pickling). The reconstructibility is key to implement the elemental function reuse technique illustrated above

Note that the developer doesn't need to do absolutely anything to exploit this feature. It gets applied automatically in between IET passes. This simplifies the developer life in all sort of ways.

This feature is currently exploited by two major compiler passes:

* MPI (which produce per-Function halo exchange elemental functions)
* asynchronous streaming (which produce per-Function data streaming elemental functions)

Most of these elemental functions are semantically identical. While MPI had a caching mechanism to catch twins, the asynchrnous streaming had no such a thing. Now with this feature we drop the caching mechanism in MPI, and both passes (well any passes) automatically exploit it. 

Upcoming PRs will also greatly benefit from this

Plus a small change in creating ipyparallel profiles
rm -rf $IPYTHONDIR/profile_mpi/

### CODE REVIEW

File: tests/test_gpu_common.py [hunk 1]
```diff
@@ -415,7 +411,7 @@ def test_streaming_basic(self, opt, ntmps):
         assert np.all(u.data[1] == 36)
 
     @pytest.mark.parametrize('opt,ntmps,nfuncs', [
-        (('buffering', 'streaming', 'orchestrate'), 4, 6),
+        (('buffering', 'streaming', 'orchestrate'), 7, 3),
```
Comment (FabioLuporini) [inline tests/test_gpu_common.py:414]: note the number of efuncs has halved !

File: devito/mpi/routines.py [hunk 2]
```diff
@@ -321,6 +318,7 @@ def _make_copy(self, f, hse, key, swap=False):
         eqns.extend([Eq(d.symbolic_max, d.symbolic_size - 1) for d in dims])
         eqns.append(eq)
 
+        # Compile `eqns` into an IET via recursive compilation
```
Comment (georgebisbas) [inline devito/mpi/routines.py:321]: Lower

File: devito/passes/iet/engine.py [hunk 3]
```diff
@@ -25,64 +28,33 @@ class Graph(object):
...
 
-            # Track all objects introduced by `func`
             self.includes.extend(as_tuple(metadata.get('includes')))
             self.headers.extend(as_tuple(metadata.get('headers')))
-            self.ffuncs.extend(as_tuple(metadata.get('ffuncs', [])))
-            self.efuncs.update(OrderedDict([(i.name, i)
-                                            for i in metadata.get('efuncs', [])]))
 
-            # Update compiler if necessary
+            # Update jit-compiler if necessary
```
Comment (georgebisbas) [inline devito/passes/iet/engine.py:57]: JIT-compiler?

File: devito/passes/iet/engine.py [hunk 4]
```diff
@@ -96,66 +68,20 @@ def apply(self, func, **kwargs):
...
-                        # Temporaries (e.g., Arrays) *cannot* be args in `root`.
-                        # So if we end up here, `a` keeps being undefined
-                        # inside it, and we rely on a later pass to define it
-                        continue
-
-                    processed.append(a)
 
-                processed = [a for a in processed if a not in drop_args]
+            # Minimize code size by abstracting semantically identical efuncs
+            efunc, efuncs = reuse_efuncs(efunc, metadata.get('efuncs', []))
```
Comment (georgebisbas) [inline devito/passes/iet/engine.py:73]: Can we avoid singular-plural for these names?

File: devito/passes/iet/engine.py [hunk 5]
```diff
@@ -202,3 +128,243 @@ def iet_visit(func):
 
 
 Jitting = namedtuple('Jitting', 'includes include_dirs libs lib_dirs')
+
+
+def create_call_graph(root, efuncs):
+    """
+    Create a Call graph -- a Direct Acyclic Graph with edges from callees
```
Comment (georgebisbas) [inline devito/passes/iet/engine.py:135]: Directed

File: devito/passes/iet/engine.py [hunk 6]
```diff
@@ -202,3 +128,243 @@ def iet_visit(func):
...
+
+
+def update_args(root, efuncs, dag):
+    """
+    Re-derive the parameters of `root` and apply the changes in cascade through
+    the `efuncs`.
+
+    For example, given the Callable `root`:
+
+        root(x, y) : x + z
```
Comment (georgebisbas) [inline devito/passes/iet/engine.py:297]: x + y

File: devito/passes/iet/engine.py [hunk 7]
```diff
@@ -202,3 +128,243 @@ def iet_visit(func):
...
+    drop_args = [a for a in root.parameters if a in defines]
+
+    if not (new_args or drop_args):
+        return efuncs
+
+    def _filter(v, efunc=None):
+        processed = list(v)
+        for a in new_args:
+            if a in processed:
+                # A child efunc trying to add a symbol alredy added by a
```
Comment (georgebisbas) [inline devito/passes/iet/engine.py:341]: already

File: devito/passes/iet/engine.py [hunk 8]
```diff
@@ -202,3 +128,243 @@ def iet_visit(func):
...
+                # sibling efunc
+                continue
+
+            if isinstance(efunc, EntryFunction) and not isinstance(a, ArgProvider):
+                # Temporaries (e.g., Arrays) *cannot* be arguments of an
+                # EntryFunction. So if we end up here, `a` remains for
```
Comment (georgebisbas) [inline devito/passes/iet/engine.py:347]: remains undefined for now, inside

File: devito/tools/abc.py [hunk 9]
```diff
@@ -90,7 +90,59 @@ def _signature(self):
...
+            * `a._rebuild() -> x(3, 5, 4)` (i.e., copy of `a`).
+            * `a._rebuild(4) -> x(4, 5, 4)`
+            * `a._rebuild(4, 7) -> x(4, 7, 4)`
+            * `a._rebuild(c=5) -> x(3, 5, 5)`
+            * `a._rebuild(1, c=7) -> x(1, 5, 7)`
+        """
+        args += tuple(getattr(self, i) for i in self.__rargs__[len(args):])
+        kwargs.update({i: getattr(self, i) for i in self.__rkwargs__ if i not in kwargs})
+
+        # Should we use a constum reconstructor?
```
Comment (georgebisbas) [inline devito/tools/abc.py:136]: constum ?

File: devito/types/sparse.py [hunk 10]
```diff
@@ -1169,6 +1165,11 @@ class MatrixSparseTimeFunction(AbstractSparseTimeFunction):
     _time_position = 0
     """Position of time index among the function indices."""
 
+    # We use DiscreteFunction instead of AbstractSparseTimeFunction
+    # because we want to get rid of 'npoint'
+    __rkwargs__ = (DiscreteFunction.__rkwargs__ +
+                   ('dimensions', 'r', 'matrix', 'nt', 'grid'))
+
     def __init_finalize__(self, *args, **kwargs):
         # The crucial argument to DugSparseTimeFunction is a sparse
```
Comment (georgebisbas) [inline devito/types/sparse.py:1174]: Dug?

File: tests/test_linearize.py [hunk 11]
```diff
@@ -169,9 +168,9 @@ def test_codegen_quality0():
     assert len(exprs) == 6
     assert all('const long' in str(i) for i in exprs[:-2])
 
-    # Only four access macros necessary, namely `uL0`, `aL0`, `bufL0`, `bufL1` (the
+    # Only four access macros necessary, namely `uL0`, `bufL0`, `bufL1` (the
```
Comment (georgebisbas) [inline tests/test_linearize.py:171]: four -> three?

**Pull request merged**